diff --git a/arch/l4/Makefile b/arch/l4/Makefile
index a63811508..e5ca479ba 100644
--- a/arch/l4/Makefile
+++ b/arch/l4/Makefile
@@ -291,10 +291,11 @@ L4LX_E_L-$(CONFIG_L4_PDSPBLK_DRV)  += pers_dsp
 L4LX_E_L-$(CONFIG_L4_TCG_STPM)     += stpm-client
 L4LX_E_L-$(CONFIG_L4_FERRET)       += ferret
 
-#L4INC		= -I$(L4OBJ)/include/$(L4_MK_ARCH)/$(L4_MK_API) \
-#		  -I$(L4OBJ)/include/$(L4_MK_ARCH) \
-#		  -I$(L4OBJ)/include/$(L4_MK_API) \
-#		  -I$(L4OBJ)/include
+###
+L4INC		= -I$(L4OBJ)/include/$(L4_MK_ARCH)/$(L4_MK_API) \
+		  -I$(L4OBJ)/include/$(L4_MK_ARCH) \
+		  -I$(L4OBJ)/include/$(L4_MK_API) \
+		  -I$(L4OBJ)/include
 
 head-y := arch/l4/kernel/head.o
 
diff --git a/arch/l4/kernel/arch-arm/vmlinux.lds.S b/arch/l4/kernel/arch-arm/vmlinux.lds.S
index ee3aaa737..3d1b95a59 100644
--- a/arch/l4/kernel/arch-arm/vmlinux.lds.S
+++ b/arch/l4/kernel/arch-arm/vmlinux.lds.S
@@ -108,6 +108,12 @@ SECTIONS
 		LONG(0xffffffff);
 		_parent_cap_local_name = .;
 		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
 	} : rw
 
 	. = ALIGN(4096);
diff --git a/arch/l4/kernel/arch-x86/vmlinux.lds.S b/arch/l4/kernel/arch-x86/vmlinux.lds.S
index 4b2b196bf..f8bfc6b60 100644
--- a/arch/l4/kernel/arch-x86/vmlinux.lds.S
+++ b/arch/l4/kernel/arch-x86/vmlinux.lds.S
@@ -101,6 +101,13 @@ SECTIONS
 		LONG(0xffffffff);
 		_parent_cap_local_name = .;
 		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
+		LONG(0xffffffff);
 
 		/*
 		 * Platform-specific entry for Fiasco.OC.
